bokehjs/src/coffee/models/widgets/dialog.coffee

```diff
@@ -65,15 +64,42 @@ class Dialog extends Widget.Model
   type: "Dialog"
   default_view: DialogView
 
+  initialize: (options) ->
+    super(options)
+    @children = []
+    if typeof(@content) != "string"
+      @children.push(@content)
+
   @define {
       visible:     [ p.Bool,    false ]
       closable:    [ p.Bool,    true  ]
       title:       [ p.String,  ""    ]
-      content:     [ p.String,  ""    ]
+      content:     [ p.Any,     ""    ]  # str or LayoutDOM
       buttons:     [ p.Array,   []    ]
       buttons_box: [ p.Instance       ]
     }
 
+  @internal {
+      children:    [ p.Array,   []    ]
+  }
+
+  get_layoutable_children: () ->
+    return @get('children')
+
+  get_edit_variables: () ->
+    edit_variables = super()
+    # Go down the children to pick up any more constraints
+    for child in @get_layoutable_children()
+      edit_variables = edit_variables.concat(child.get_edit_variables())
+    return edit_variables
+
+  get_constraints: () ->
+    constraints = super()
+    # Go down the children to pick up any more constraints
+    for child in @get_layoutable_children()
+      constraints = constraints.concat(child.get_constraints())
+    return constraints
```

Comment:
I agree that a  separate solver is probably ideal. It will take some investigation to see if this is straightforward or now. I think it may not be so bad, we had already discussed a path for per-root solvers.